kong/pdk/client.lua

```diff
@@ -199,6 +200,63 @@ local function new(self)
   end
 
 
+  ---
+  -- Returns the Nginx subsystem used by the client. It can be
+  -- "http" or "stream"
+  --
+  -- @function kong.client.get_subsystem
+  -- @phases access, header_filter, body_filter, log
+  -- @treturn string a string with either `"http"` or `"stream"`
+  -- @usage
+  -- kong.client.get_subsystem() -- "http"
+  function _CLIENT.get_subsystem()
+    check_phase(AUTH_AND_LATER)
+
+    return ngx.config.subsystem
+  end
```

Comment:
Which subsystem is in use is an implementation detail of Nginx, not a client's choice (unlike choosing a protocol). It makes sense that `get_protocol()` be in this module, but `get_subsystem()` should be a top-level API: `kong.get_subsystem()` answers the question "which Nginx subsystem is my code currently executing within?"

Reply:
I thought some of the PDK was meant to be decoupling from nginx?
It feels like an implementation detail that http and stream workers aren't shared.

Reply:
That is precisely my point.

Reply:
I don't follow your argument: shouldn't code be asking "what is the relevant request type for the current client?" rather than "what subsystem is this code running in?"
The former is always a sensible question to ask; the latter only makes sense as long as there are different lua VMs for http vs stream.

Reply:
> what is the relevant request type for the current client?

get_protocol() should answer this question.

However, the subsystem isn’t tied to a client, but is an nginx/OpenResty implementation detail. It is necessary to know the current subsystem in contexts where’s no clients are connected. See my other comment about this.

This necessity for 2 APIs (protocol _and_ subsystem) was discussed during a Core Team spoke a couple of months ago. Ultimately subsystem should be less and less relevant. It is 11PM on a Sunday, so forgive the brevity here.